tools/keyvaluestore.go

```diff
@@ -0,0 +1,201 @@
+package tools
+
+import (
+	"encoding/gob"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+// KeyValueStore provides an in-memory key/value store which is persisted to
+// a file. The file handle itself is not kept locked for the duration; it is
+// only locked during load and save, to make it concurrency friendly. When
+// saving, the store uses optimistic locking to determine whether the db on disk
+// has been modified by another process; in which case it loads the latest
+// version and re-applies modifications made during this session. This means
+// the Lost Update db concurrency issue is possible; so don't use this if you
+// need more DB integrity than Read Committed isolation levels.
+type KeyValueStore struct {
+	mu       sync.RWMutex
+	filename string
+	log      []keyValueChange
+
+	// This is the persistent data
+	// version for optimistic locking, this field is incremented with every Save()
+	version int64
+	db      map[string]interface{}
+}
+
+type keyValueOperation int
+
+const (
+	// Set a value for a key
+	keyValueSetOperation = keyValueOperation(iota)
+	// Removed a value for a key
+	keyValueRemoveOperation = keyValueOperation(iota)
+)
+
+type keyValueChange struct {
+	operation keyValueOperation
+	key       string
+	value     interface{}
+}
+
+// NewKeyValueStore creates a new store and initialises it with contents from
+// the named file, if it exists
+func NewKeyValueStore(filepath string) (*KeyValueStore, error) {
+	kv := &KeyValueStore{filename: filepath, db: make(map[string]interface{})}
+	err := kv.loadAndMergeIfNeeded()
+	return kv, err
+}
+
+// Set updates the key/value store in memory
+// Changes are not persisted until you call Save()
+func (k *KeyValueStore) Set(key string, value interface{}) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	k.db[key] = value
+	k.log = append(k.log, keyValueChange{keyValueSetOperation, key, value})
+}
+
+// Remove removes the key and its value from the store in memory
+// Changes are not persisted until you call Save()
+func (k *KeyValueStore) Remove(key string) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	delete(k.db, key)
+	k.log = append(k.log, keyValueChange{keyValueRemoveOperation, key, nil})
```

Comment:
Looks like these transactional methods are pretty similar. What do you think about teaching the `keyValueChange` type how to perform these operations?